engine/runtime/src/main/scala/org/enso/compiler/Compiler.scala

```diff
@@ -218,7 +218,23 @@ class Compiler(
     initialize()
     modules.foreach(m => parseModule(m))
 
-    var requiredModules = modules.flatMap(runImportsAndExportsResolution)
+    var requiredModules = modules.flatMap { module =>
+      val modules = runImportsAndExportsResolution(module)
+      if (
+        module
+          .wasLoadedFromCache() && modules.exists(!_.wasLoadedFromCache())
```

Comment:
Does the `modules.exists` check just direct dependencies of a module? What if there is a transitive dependency that is invalidated? Three modules depends one on another `A -> B ->C`. `C` caches are out of date. When `A` module is being checked it is OK and `B` is also OK. But then `B` is checked and gets invalidated because of `C`.

Will such situation be handled correctly?

Reply:
Yes, it will. Thanks to the way the current import/export resolution is handled.
I double checked :)